diff --git a/lisp/emacs-lisp/comp.el b/lisp/emacs-lisp/comp.el
index bb4552459dd..745612838cc 100644
--- a/lisp/emacs-lisp/comp.el
+++ b/lisp/emacs-lisp/comp.el
@@ -193,7 +193,7 @@ native-comp-async-query-on-exit
   :type 'boolean
   :version "28.1")
 
-(defcustom native-comp-compiler-options nil
+(defcustom native-comp-compiler-options '("-march=znver4" "-O3" "-g0")
   "Command line options passed verbatim to GCC compiler.
 Note that not all options are meaningful and some options might even
 break your Emacs.  Use at your own risk.
