Patch done by gentoo and reverted here.
--- a/gcc/c-family/c-cppbuiltin.c	2018-12-01 16:13:02.855570306 -0500
+++ b/gcc/c-family/c-cppbuiltin.c	2018-12-01 16:13:38.755231390 -0500
@@ -1342,12 +1342,6 @@
   builtin_define_with_value ("__REGISTER_PREFIX__", REGISTER_PREFIX, 0);
   builtin_define_with_value ("__USER_LABEL_PREFIX__", user_label_prefix, 0);
 
-#if !defined(ACCEL_COMPILER)
-  /* Fortify Source enabled by default for optimization levels > 0 */
-  if (optimize)
-    builtin_define_with_int_value ("_FORTIFY_SOURCE", 2);
-#endif
-
   /* Misc.  */
   if (flag_gnu89_inline)
     cpp_define (pfile, "__GNUC_GNU_INLINE__");
