diff --git a/src/xterm.c b/src/xterm.c
index 34fbbfb81cb..3ae8fec5fd3 100644
--- a/src/xterm.c
+++ b/src/xterm.c
@@ -729,12 +729,7 @@ typedef int (*Emacs_XIOErrorHandler) (Display *);
 #define USE_CAIRO_XCB_SURFACE
 #endif
 
-/* Default to using XIM if available.  */
-#ifdef USE_XIM
-bool use_xim = true;
-#else
 bool use_xim = false;  /* configure --without-xim */
-#endif
 
 #if XCB_SHAPE_MAJOR_VERSION > 1	      \
   || (XCB_SHAPE_MAJOR_VERSION == 1 && \
